File: leetcode/92.reverse-linked-list-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        
        if (m >= n)
        {
            return head;
        }

        ListNode newHead(0);
        newHead.next = head;
        ListNode *cur = &newHead;

        for (auto i = 0; i < m - 1; ++i)
        {
            cur = cur->next;
        }

        head = cur;
        cur = cur->next;

        for (auto i = 0; i < n - m; ++i)
        {
            ListNode *tmp = cur->next;
            cur->next = cur->next->next;

            tmp->next = head->next;
            head->next = tmp;
        }

        return newHead.next;
    }
};

File: leetcode/92.reverse-linked-list-ii.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseBetween(self, head, m, n):
        """
        :type head: ListNode
        :type m: int
        :type n: int
        :rtype: ListNode
        """

        new_head = ListNode(0)
        new_head.next = head

        # find node.next = target
        cur = new_head
        for _ in range(m - 1):
            cur = cur.next

        head = cur
        cur = cur.next

        # n1, n2... nk, move n2, n3... nk before n1
        # times = n - 1
        # loop time = (n - m + 1) - 1
        for _ in range(n - m):

            node = cur.next
            cur.next = cur.next.next

            node.next = head.next
            head.next = node
            

        return new_head.next
